ipq806x: various fixes and improvements

Ansuel opened the pull request:
This pr contains various fix for the ipq806x platform
This pr contains: 
- It was notice that there is a problem with L2 cache scaling. This fixes that. (It has been proved by various user that have reported increased performance and stability)
- Update the cpufreq driver to use opp table and use a more standard way. Detailed description in commits message
- Rework tsense drive to fix a wrong error output and mem leak

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: fix bug in L2 cache scaling

It has been notice a buf in L2 cache scaling where the scaling is not
done proprely if the frequency is set to the initial state before
the new frequency.

From: https://patchwork.kernel.org/patch/10565443/

* The clocks are set to aux clock rate first to make sure the
* secondary mux is not sourcing off of QSB. The rate is then set to
* two different rates to force a HFPLL reinit under all
* circumstances.

In the initial stage of boot to force a new frequency to apply, is
needed to first set the frequency back to the lowest one (aux_rate)
and then to the target one. This force and make sure the controller
actually switch the frequency to the right one. Apply the same
mechanism to L2 frequency scaling. Before scaling to the target
frequency, first set the frequency to the aux_rate to force the
transition, then scale it to the target frequency. Doing the wrong way
can produce unexpected results and could lock the scaling mechanism
until a full reboot is done (Causing a full reset by the krait-cc driver)

From: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=77612720a2362230af726baa4149c40ec7a7fb05

When the Hfplls are reprogrammed during the rate change,
the primary muxes which are sourced from the same hfpll
for higher frequencies, needs to be switched to the 'safe
secondary mux' as the parent for that small window. This
is done by registering a clk notifier for the muxes and
switching to the safe parent in the PRE_RATE_CHANGE notifier
and back to the original parent in the POST_RATE_CHANGE notifier.

This should apply also to L2 scaling... as we can't relly use
the notifier, we manually do this on L2 scaling.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: opp/core: update patch allowing adjusting of OPP voltages at runtime

Update 0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime with
upstream version.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: backport qcom-cpufreq-kryo patch

Backport patch applied to qcom-cpufreq-kryo
driver as krait cpu will base on this driver.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: add patch to support krait cpu scaling driver

This patch has been proposed but never actually merged to
mainline. It was accepted but never re proposed by the
creator.
Rework it, fix kernel panic cause by double kfree.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: drop old cpufreq patch

Drop old cpufreq as now we have new driver that
can use normal kernel opp definition

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: opp/core: update patch updating voltage tolerance on voltage adjust

Rework 0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t
to reflect changes upstream.

- Skip unnecessary allocation of buffer to set u_volt
- Change opp u_volt directly

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: refresh new and changed patches

Refresh patches to remove fuzz
Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: update kernel config with new driver

The config name for cpufreq driver has changed.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: rework ipq806x specific tsense temp driver

Tsense driver for ipq806x have various problem.
- Emit wrong error. On probing of this driver, nvmem driver can be
  not ready and this cause a EDEFER error. Actually this is not an
  error as the kernel will retry to probe the driver after the
  nvmem driver is loaded.
- Use uninitialized value on trigger of critical temp
- Doesn't free allocated memory

Because of this, rework the driver and improve it by removing extra
load of data.

Change the logic of loading data. Use the backup calib data only
when the calib data is not present. As the calibration is only
needed to set the temp offset, we don't really need to read
both calib data and set the offset based only on the backup one.
Also change how the notifier function work. At times when we
output the trigger message, we already have read the temp so
remove the extra read and the wrong uninitialized data that
probably caused a kernel panic for null pointer exception.
(Think we never experience this bug because the router
never reached that temp ever... So just lucky)

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@ynezz sorry if i bother you again...  can you review this?

pkgadd commented:
I've been testing this for the last ~20 hours successfully on my nbg6817/ ipq8065.

`Tested-by: Stefan Lippers-Hollmann <s.l-h@gmx.de>`

Ansuel commented:
I use this for 4 days without a problem 
I also received a private email that reported me that this fix a stability problem with some app (Tvheadend 4.3) that cause freeze and reboot on both kernel 4.14 and 4.19

neheb commented:
I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

hnyman commented:
Some ideas:
* The L2 scaling issue commit message might contain a reference to the upstream commit/discussion (codeaurora(?)), forum discussion(?), from where the idea of changing the cache speed via the lowest speed surfaced.

* The tsense driver changes (thermal statistics) commit messages might have an explanation that the changes concern the thermal statistics. "tsense" is rather cryptic.

Ansuel commented:
i really can't find now the commit but if i remember correctly... we find that a fix was merged to upstream related to another arm cpu that had the exact same problem. (Problem transitioning from one frequency to another and the necessity to revert to idle frequency before jumping to the new one)

About t sense... updated...

hnyman reviewed (commented):
Nitpicking, but it should be "kryo", not "kyro" in the commit title

SqTER-PL commented:
> I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

I use this compilation over 3 days very intensively. Works better like a Master on 4.14 kernel. I didn't notice throughput and latency issues. NAT and USB work faster. Zero errors, very stable.

SqTER-PL commented:
> I'm wondering if this improves ethernet support. Last I used my R7800 I was having throughput and latency issues.

I use this compilation on R7800 over 2 days very intensively. Works better like a Master on 4.14 kernel. I didn't notice throughput and latency issues. NAT and USB work faster. Zero errors, very stable.
@Ansuel - good job 👍

Ansuel commented:
@hnyman i finally find the patch i was talking about... Added an extensive explanation about the L2 fix.

shelterx commented:
Works great here too.

Ansuel commented:
News on this?

SqTER-PL commented:
It works very well, which is good news.

ynezz commented on target/linux/ipq806x/patches-4.19/0052-PM-OPP-Update-the-voltage-tolerance-when-adjusting-t.patch:
```diff
@@ -19,20 +19,20 @@ Signed-off-by: Georgi Djakov <georgi.djakov@linaro.org>
 +++ b/drivers/opp/core.c
```
This commit subject and description needs more love.

Ansuel replied:
Good now?

ynezz replied:
I've meant this one ![image](https://user-images.githubusercontent.com/43848/71447903-24100100-2735-11ea-8748-bd92777a1c73.png) as it's not possible to comment on it...

Ansuel replied:
Sorry for the stupid question... How should i change them ?

ynezz replied:
Imagine, that you're reading the Git history somewhere in 2021 and see `ipq806x: update 0052`, would you be able to tell from this commit subject what is this commit about?

ynezz replied:
And this patch is probably some upstream backport, so perhaps your changes should be upstreamed first as well?

Ansuel replied:
Thx for the explanation. 
About proposing the patch upstream i can try... Anyway i have updated it for now...

ynezz commented on target/linux/ipq806x/patches-4.19/0049-PM-OPP-Support-adjusting-OPP-voltages-at-runtime.patch:
```diff
@@ -1,7 +1,14 @@
-From c949f08cf20fe82971fbdb4015daa38210da492e Mon Sep 17 00:00:00 2001
```
Commit subject/description should be improved.

Ansuel replied:
Is it good now? I put the merged patch

ynezz replied:
I've meant this one ![image](https://user-images.githubusercontent.com/43848/71447899-06429c00-2735-11ea-9760-534efc2d5909.png) as it's not possible to comment on this...

ynezz replied:
Perhaps something like this?

`ipq806x: opp/core: update patch allowing adjusting of OPP voltages at runtime`

ynezz commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi:
```diff
@@ -84,3 +31,88 @@
+	compatible = "operating-points-v2-qcom-cpu";
+	nvmem-cells = <&speedbin_efuse>;
+
+	// On ipq8065 1.2 ghz freq is not present
```
I would remove the comment and added it to the commit description, if you insist on the comment, then please use /**/ format.

Ansuel replied:
Think is better to put a comment in the dtss

ynezz commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi:
```diff
@@ -84,3 +31,88 @@
+		opp-microvolt-speed0-pvs6-v0 = <775000>;
+		opp-supported-hw = <0x1>;
+		clock-latency-ns = <100000>;
+	};
```
What about newline between nodes?

ynezz commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8065.dtsi:
```diff
@@ -84,3 +31,88 @@
+		opp-supported-hw = <0x1>;
+		clock-latency-ns = <100000>;
+	};
+
```
This newline is not needed.

ynezz commented on target/linux/ipq806x/patches-4.19/0063-4-ip806x-tsense-rework-driver.patch:
```diff
@@ -0,0 +1,107 @@
+--- a/drivers/thermal/qcom/tsens-ipq8064.c
```
Line wrap the commit description to < 72 chars.

ynezz commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi:
```diff
@@ -573,10 +573,10 @@
 			#size-cells = <1>;
```
`11 sensor` -> `11 sensors`
`reg propriety` -> `reg property`

ynezz commented on target/linux/ipq806x/files-4.19/arch/arm/boot/dts/qcom-ipq8064.dtsi:
```diff
@@ -573,10 +573,10 @@
 			};
 			tsens_backup: backup@410 {
-				reg = <0x410 0x10>;
+				reg = <0x410 11>;
```
What about using hex notation?

Ansuel replied:
Changed to hex but i still think 11 better explain it...

If we change to hex here should we change also speedbin to 0x4 ?

ynezz commented:
> Some ideas:
> 
>     * The tsense driver changes (thermal statistics) commit messages might have an explanation that the changes concern the thermal statistics. "tsense" is rather cryptic.

+1

ynezz commented:
> It works very well, which is good news.

@SqTER-PL Which device?

Ansuel committed:
ipq8065: update dtsi with new opp table

The new driver use opp table to register frequency.
Drop psv bindings as they are not used anymore.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: fix wrong definition for tsense calib

We have only 11 sensors on ipq806x. Fix the reg property
to load the right amount of data instead of the entire
space.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq8064: update dtsi with new opp table

The new driver use opp table to register frequency.
Drop psv bindings as they are not used anymore.
Adds speedbin definition for nvmem driver

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@ynezz everything should be good except for the voltage tolerance patch

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
